controllers/bmcjob_controller.go

```diff
@@ -18,38 +18,207 @@ package controllers
 
 	bmcv1alpha1 "github.com/tinkerbell/rufio/api/v1alpha1"
 )
 
 // BMCJobReconciler reconciles a BMCJob object
 type BMCJobReconciler struct {
-	client.Client
-	Scheme *runtime.Scheme
+	client           client.Client
+	bmcClientFactory BMCClientFactoryFunc
+	bmcTaskHandler   BMCTaskHandler
+	logger           logr.Logger
+}
+
+// NewBMCJobReconciler returns a new BMCJobReconciler
+func NewBMCJobReconciler(client client.Client, bmcClientFactory BMCClientFactoryFunc, bmcTaskHandler BMCTaskHandler, logger logr.Logger) *BMCJobReconciler {
+	return &BMCJobReconciler{
+		client:           client,
+		bmcClientFactory: bmcClientFactory,
+		bmcTaskHandler:   bmcTaskHandler,
+		logger:           logger,
+	}
 }
 
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the BMCJob object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+// Reconcile runs a BMCJob.
+// Creates the individual BMCTasks on the cluster.
+// Runs each Task in a BMCTask and updates the BMCJob status.
 func (r *BMCJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := r.logger.WithValues("BMCJob", req.NamespacedName)
+	logger.Info("Reconciling BMCJob")
+
+	// Fetch the bmcJob object
+	bmcJob := &bmcv1alpha1.BMCJob{}
+	err := r.client.Get(ctx, req.NamespacedName, bmcJob)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Failed to get BMCJob")
+		return ctrl.Result{}, err
+	}
+
+	// Deletion is a noop.
+	if !bmcJob.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	// Job has StartTime is noop.
+	if !bmcJob.Status.StartTime.IsZero() {
+		return ctrl.Result{}, nil
+	}
+	// Create a patch from the initial BMCJob object
+	// Patch is used to update Status after reconciliation
+	bmcJobPatch := client.MergeFrom(bmcJob.DeepCopy())
+
+	return r.reconcile(ctx, bmcJob, bmcJobPatch, logger)
+}
+
+func (r *BMCJobReconciler) reconcile(ctx context.Context, bmj *bmcv1alpha1.BMCJob, bmjPatch client.Patch, logger logr.Logger) (ctrl.Result, error) {
+	// Get BaseboardManagement object for the Job
+	// Requeue if error
+	baseboardManagement := &bmcv1alpha1.BaseboardManagement{}
+	err := r.resolveBaseboardManagementRef(ctx, bmj, baseboardManagement)
```

Comment:
Re the method name, `resolveBaseboardManagementRef`, I don't think we're resolving anything. There's a 1 to 1 mapping of reference to object so we're simply doing a retrieval operation. 'Resolve' implies something else (naming is hard).

Reply:
Fair point. I can rename this to something that describes the operation which like you said is a `Get`.